Add REST store support

lzhoucs opened the pull request:
**Type:** feature

**Description:** 
Add a REST based storage that interacts with a target REST endpoint remotely. Refactored `createInMemoryStorage`

**Related Issue:** #47 , #4 

Please review this checklist before submitting your PR:

* [x] There is a related issue
* [x] All contributors have signed a CLA
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [x] The code passes the CI tests
* [x] Unit or Functional tests are included in the PR
* [x] The PR increases or maintains the overall unit test coverage percentage
* [x] The code is ready to be merged

lzhoucs committed:
Rest storage implementation

maier49 commented on src/storage/createInMemoryStorage.ts:
```diff
@@ -1,72 +1,26 @@
-			});
-		}
-	},
+const createInMemoryStorage: InMemoryStorageFactory = createBaseStorage.mixin(compose({
```
If we're using `mixin()`, is there a reason we're also using `compose()`, or could we just pass a mixin descriptor? e.g. 

``` Typescript
{
  mixin: {
    createId(...)
    ...
  },
  initialize(instance, options) {
  }
}
```

lzhoucs replied:
Yeah. That's a good idea and I tried it. But it doesn't seem to work. You can find the change in [this branch](https://github.com/lzhoucs/stores/blob/rest-storage-feedback/src/storage/createRestStorage.ts#L76)

tomdye replied:
Does seem like a strange usage, should work with the format bradley used above

maier49 replied:
This is now using a mixin descriptor argument instead of a factory.

lzhoucs commented:
All tests passed. The build is hanging on "Created session iphone 9.3 on any platform". It doesn't seem to be caused by my changes.

maier49 commented on src/storage/createRestStorage.ts:
```diff
@@ -0,0 +1,283 @@
+			uri += '/' + ids[0];
+		}
+		else {
+			uri += '?' + state.idSerializer(ids);
```
We should probably check if `state.idSerializer` exists and default to making individual get requests here if it doesn't.

lzhoucs replied:
Thanks. This makes sense to me. I will make the change here(`get`) as well as in `delete`

maier49 commented on src/storage/createRestStorage.ts:
```diff
@@ -0,0 +1,283 @@
+			uri += '/' + ids[0];
+		}
+
+		return <Promise<any>> request(uri, {
```
We should probably check if `state.requestSerializer` exists and default to making individual put requests here if it doesn't.

lzhoucs replied:
This one is tricky. `requestSerializer` is used as the `object => string` serializer regardless of the type of the request. So it can't be used as an indicator of opt in custom request. I can probably add another option just for `put` custom request scenario, unless you have other ideas.

maier49 commented on src/storage/createRestStorage.ts:
```diff
@@ -0,0 +1,283 @@
+	}
+}, <T, O>(instance: Storage<T, O>, {target,
+									createIdTarget,
+									idSerializer = defaultIdSerializer,
```
The serializers used for multiple puts and multiple gets probably shouldn't be defaulted. If they are supporting non standard ways of doing things as a convenience we will probably want to make that an opt in behavior.

lzhoucs replied:
I agree. I am making changes for this. For `get` and `delete`, I will repurpose `idSerializer` as the indicator, and for `put`, I will add another option, since `requestSerializer` is being used in multiple scenarios.

lzhoucs commented:
I made the discussed changes. Will do a rebase once other PRs are landed.

tomdye commented on src/storage/createBaseStorage.ts:
```diff
@@ -0,0 +1,74 @@
+			return itemArray.map(state.idFunction);
+		}
+		else {
+			return itemArray.map(function(item) {
```
`return itemArray.map((item) => item.id);`

tomdye commented on src/storage/createBaseStorage.ts:
```diff
@@ -0,0 +1,74 @@
+		const state = instanceStateMap.get(this);
+		const itemArray = Array.isArray(items) ? <IdObject []> items : [ <IdObject> items ];
+		if (state.idProperty) {
+			return itemArray.map((item) => {
```
`return itemArray.map((item) => item[ state.idProperty ]);`

tomdye commented on src/storage/createRestStorage.ts:
```diff
@@ -0,0 +1,297 @@
+				headers: {
+					'Accept': state.contentType
+				}
+			}).then(function(response) {
```
utilise the arrow functions:
`.then((response) => state.responseDeserializer(response.data.toString()));`

tomdye commented on tests/unit/storage/createRestStorage.ts:
```diff
@@ -0,0 +1,812 @@
+	},
+
+	'GET': {
+		'ids can\'t be empty'() {
```
Can you be a bit more descriptive with your names please. ie.
`Should throw error when get called without ids`

tomdye commented:
Have only done a brief review.
Also why the usage of `String(thing)` all over the place over `thing.toString()`?

lzhoucs commented:
Thanks @Tomdye for the review. I addressed the feedback.

Regarding the `String` over `toString()`, ideally none of them should be used since `response.data` is already a `string`. However there were some typing issues, so I used the `String` wrapper to get around the issue in 3 places. Anyway my latest commit have addressed them.

tomdye commented:
@lzhoucs I would imagine the `typings issues` you described are just that, and should be fixed via the typings? I'm probably wrong though. As a general rule we should avoid using `any` as much as possible.

Tomdye commented on src/storage/createRestStorage.ts:
```diff
@@ -91,13 +89,11 @@ const createRestStorage: RestStorageFactory = createBaseStorage.mixin(compose({
 			uri += '?' + query.toString(state.querySerializer);
 		}
-		return <Promise<any>> request.get(uri, {
+		return <Promise<any>> request.get<string>(uri, {
```
I *think* this should be `<Promise<string>>`

lzhoucs commented:
You're right @Tomdye. And I wouldn't be surprised if typing issues were fixed by typing changes. As you can see from [my change](https://github.com/dojo/stores/pull/48/commits/aba340eef651e8e45ef6bdf717a9193ebcd45beb#diff-1641750974acc5362db33cc6110f5339R92), I added generic typing `<string>` when calling `request.get` so `response.data` will be `string` as expected, as such there's no need to do any type of conversion as I did with `String()` wrapper.

As for `any`, I did try not to use it as much as possible. If there's any place you think I can replace `any` with a concrete type, please let me know and I am happy to make the change.

dylans commented:
This one was ready to land, but we're waiting for https://github.com/dojo/core/pull/261 to land and then we'll refactor this on top of that.

dylans reviewed (commented):
This will likely need some refactoring when https://github.com/dojo/core/pull/261 lands.

maier49 commented:
Closing in favor of #124 which has been rebased against the latest from master and works with the updated request API from `dojo/core`
